File: javascript/src/AstNode.ts
import type IToken from './IToken'
import type { RuleType, TokenType } from './Parser'

export default class AstNode {
  // biome-ignore lint/suspicious/noExplicitAny: keys and values are heterogeneous AST entries
  private readonly subItems = new Map<any, any[]>()

  constructor(public readonly ruleType: RuleType) {}

  // biome-ignore lint/suspicious/noExplicitAny: keys and values are heterogeneous AST entries
  public add(type: any, obj: any) {
    let items = this.subItems.get(type)
    if (items === undefined) {
      items = []
      this.subItems.set(type, items)
    }
    items.push(obj)
  }

  public getSingle(ruleType: RuleType) {
    return (this.subItems.get(ruleType) || [])[0]
  }

  public getItems(ruleType: RuleType) {
    return this.subItems.get(ruleType) || []
  }

  public getToken(tokenType: TokenType) {
    return (this.subItems.get(tokenType) || [])[0]
  }

  public getTokens(tokenType: TokenType): IToken<TokenType>[] {
    return this.subItems.get(tokenType) || []
  }
}

File: dart/lib/src/ast_node.dart
import 'package:cucumber_gherkin/src/language/token.dart';
import 'package:cucumber_gherkin/src/parser/parser.g.dart';

class AstNode {
  AstNode(this.ruleType);

  final Map<RuleType, List<Object?>> _subItems = <RuleType, List<Object?>>{};

  final RuleType ruleType;

  void add(RuleType ruleType, Object? obj) {
    (_subItems[ruleType] ??= <Object?>[]).add(obj);
  }

  T? getSingle<T>(RuleType ruleType) {
    final list = _subItems[ruleType];
    if (list == null || list.isEmpty) {
      return null;
    }
    return list.first as T;
  }

  List<T> getItems<T>(RuleType ruleType) {
    final list = _subItems[ruleType];
    if (list == null) {
      return <T>[];
    }
    return list.cast<T>();
  }

  Token? getToken(TokenType tokenType) => getSingle<Token>(tokenType.ruleType);

  List<Token> getTokens(TokenType tokenType) =>
      getItems<Token>(tokenType.ruleType);
}
